### ISSUE TITLE
Allow users to delete their accounts

### ISSUE THREAD

**Issue by steveoh** [uic-utah/uic-inventory#199]
Privacy management
-----------------------------------
(edited by @nathankota on Oct. 4, 2023)

Change the verbiage in the Danger Zone of the Profile page, and make the account deletion logic match:

> The following information associated with your account will be retained according to retention schedule [81505 - Underground injection control program files](https://axaemarchives.utah.gov/cgi-bin/appxretcget.cgi?WEBINPUT_RUNWHAT=HTML_1SERIES&WEBINPUT_BIBLGRPC_RID=81505&A=B):
> 
> - All submitted sites/well inventories and associated data will be retained for the retention period including: 
>   - Site name, location, and ownership 
>   - Name, email, phone number, organization, and address of associated site contacts/signatories
>   - Type, details, operating status, and location of injection wells 
> 
> The following information associated with your account will be deleted:
> 
> - All sites (including all associated information) with no well inventories
> - All sites (including all associated information) with only draft inventories
> - All draft well inventories
> - All profile information

**Comment by nathankota** [comment]
@rsparker-utah let's define the rules and requirements for this.

For example:
- Draft Inventories will be deleted
- Submitted Inventories will be deleted
- Site will be deleted if child inventories are only Draft and/or Submitted
- Approved Inventories and associated Site will _**not**_ be deleted
- ...

**Comment by rsparker-utah** [comment]
@nathankota I like the rules you’ve proposed. To add to/verify the rules you've listed when a user deletes their account:

**For Sites**
- Sites with no well inventories will be deleted
- Sites with only draft well inventories will be deleted

**For Well Inventories**
- Draft well inventories will be deleted
- Submitted (unapproved) well inventories will be deleted 
- Approved Inventories and associated Site will **_not_** be deleted

In addition to the above, there should be a note explaining to the user that the above will happen if they delete their account.

**Comment by rsparker-utah** [comment]
@nathankota Could we update the red text to something similar to the following:

The following information associated with your account will be **retained**:
- All submitted (approved) sites/well inventories 
- All essential profile information

The following information associated with your account will be **deleted**:
- All sites with no well inventories
- All sites with only draft/submitted inventories
- All draft well inventories
- All submitted (unapproved) well inventories
- All non-essential profile information

**Comment by nathankota** [comment]
Change deletion notification to Ryan's verbiage:

The following information associated with your account will be retained:
- All submitted (approved) sites/well inventories
- All essential profile information

The following information associated with your account will be deleted:
- All sites with no well inventories
- All sites with only draft/submitted inventories
- All draft well inventories
- All submitted (unapproved) well inventories
- All non-essential profile information

**Comment by nathankota** [comment]
@rsparker-utah let's think through this more, now that we know about the data retention and grama requirements, in conjunction with how we defined the approval and post-approval print functionality, specifically related to the inclusion of a submitters information in the ABR letter. 🤔

**Comment by rsparker-utah** [comment]
@nathankota Here is an updated notice based on recent GRAMA discussions:

The following information associated with your account will be retained according to retention schedule [81505 - Underground injection control program files](https://axaemarchives.utah.gov/cgi-bin/appxretcget.cgi?WEBINPUT_RUNWHAT=HTML_1SERIES&WEBINPUT_BIBLGRPC_RID=81505&A=B):

- All submitted sites/well inventories and associated data will be retained for the retention period including: 
  - Site name, location, and ownership 
  - Name, email, phone number, organization, and address of associated site contacts/signatories
  - Type, details, operating status, and location of injection wells 

The following information associated with your account will be deleted after the retention period is up:

- All sites with no well inventories
- All sites with only draft inventories
- All draft well inventories
- All profile information (i.e., name, email, phone number, organization, and address) not associated with any submitted sites/well inventories.

**Comment by steveoh** [comment]
The text says the following will be deleted after the retention period is up which is 5 years. I was under the assumption that the bottom list of items would be deleted immediately. Does this text need to be revised?

> The following information associated with your account will be deleted after the retention period is up:

**Comment by steveoh** [comment]
@rsparker-utah and @nathankota The text says the following will be deleted after the retention period is up which is 5 years. I was under the assumption that the bottom list of items would be deleted immediately. Does this text need to be revised?

> The following information associated with your account will be deleted after the retention period is up:

**Comment by nathankota** [comment]
@steveoh please change the text to:

> The following information associated with your account will be deleted:

And proceed with the logic to delete the data that is mentioned.

> 1. All sites with no well inventories
> 2. All sites with only draft inventories
> 3. All draft well inventories
> 4. All profile information (i.e., name, email, phone number, organization, and address) not associated with any submitted sites/well inventories.

**Comment by steveoh** [comment]
Based on the relationships in the database, all the contacts for sites are also deleted. I'm not sure if you want to add that to the text.

**Comment by nathankota** [comment]
> Based on the relationships in the database, all the contacts for sites are also deleted. I'm not sure if you want to add that to the text.

@rsparker-utah, this would be for unapproved sites. Do you want to add anything specific about contacts being deleted, or do you think that is understood well enough within the current text?



> The following information associated with your account will be deleted after the retention period is up:
> 
> All sites with no well inventories
> All sites with only draft inventories
> All draft well inventories
> All profile information (i.e., name, email, phone number, organization, and address) not associated with any submitted sites/well inventories.
> 
>

**Comment by steveoh** [comment]
We delete all of the profile information since they duplicate that to the contacts table for sites. Should that wording be updated?

**Comment by rsparker-utah** [comment]
Ah gotcha, how about if we trim off the "not associated with any submitted sites/well inventories." piece so it reads: 

The following information associated with your account will be deleted:

- All sites (including all associated information) with no well inventories
- All sites (including all associated information) with only draft inventories
- All draft well inventories
- All profile information (i.e., name, email, phone number, organization, and address)

**Comment by steveoh** [comment]
We don't give examples of any other information so how about we drop that also?

**Comment by rsparker-utah** [comment]
Good point. How about something along the lines of:

The following information associated with your account will be deleted:

- All sites with no well inventories
- All sites with only draft inventories
- All draft well inventories
- All profile information (i.e., name, email, phone number, organization, and address)

**Comment by steveoh** [comment]
I was more so mentioning the i.e. check the app and see what you think.

**Comment by nathankota** [comment]
Simple, clear, I think it's good. @rsparker-utah please confirm that you're okay with this.

![image](https://github.com/uic-utah/uic-inventory/assets/15659642/fb96a389-1a4e-4611-81ae-598f7be2f8f8)

**Comment by rsparker-utah** [comment]
@nathankota This looks good to me and it looks good in the web app, thanks!

**Comment by steveoh** [comment]
Yep. Ok, we can't delete the profile entirely.

> update or delete on table \"accounts\" violates foreign key constraint \"site_to_account_fk\" on table \"sites\"

**Comment by nathankota** [comment]
First name: N
Last name: K
Email: kotanl@hotmail.com

**Comment by nathankota** [comment]
My testing of the delete account feature was successful. @rsparker-utah, please validate that this works as expected, and close the issue if you are satisfied.

**Comment by rsparker-utah** [comment]
@nathankota I ran through the delete account process using one of my test accounts. The account deleted fine (i.e., it was removed from the admin user "Accounts" list) and all sites with only draft wells were also deleted as expected; however, it looks like the delete process is not currently removing draft inventories associated with sites that also have a submitted inventory. For example, the account I was testing the delete on had the following two sites associated with it prior to going through the delete process:
![image](https://github.com/uic-utah/uic-inventory/assets/49735267/32abd124-6682-42c8-8d45-a79a764e84de)
After going through the account delete process, the "231114_Test Account Delete Site" and its associated draft inventory were deleted as expected; however, the draft inventory associated with the other "Utah Lake Island Paradise TWO" site was still visible when viewing this site via my admin account:
![image](https://github.com/uic-utah/uic-inventory/assets/49735267/21193bf5-a08e-499f-aa8d-df31cde50e5f)

**Comment by steveoh** [comment]
That inventory was created with your admin user so it wasn't counted as a part of your test users inventories.

**Comment by rsparker-utah** [comment]
Ah ok. I tried some additional testing by reactivating the test account I had used. I created two new sites, one site with only a draft inventory and one site that had both a draft and submitted inventory:
![image](https://github.com/uic-utah/uic-inventory/assets/49735267/e1b4e392-37d8-4e9f-a9f5-27f4d762e15a)
I then went through the account delete process again and the site with only a draft inventory (i.e., 231115_Test Delete 2) was deleted successfully. The site that had both a draft and submitted inventory also functioned as expected with the draft inventory being deleted, but the site and submitted inventory remained as expected and is viewable via my admin account:
![image](https://github.com/uic-utah/uic-inventory/assets/49735267/58ae63cf-b376-45ca-b8a9-d4ea26131895)
The delete functionality is working as expected.
